src/economy/economy.cpp

```diff
@@ -13,33 +13,591 @@
+	if(d == 0.f) {
+		return 0.f;
+	}
+	return con / d;
+}
+float demand_satisfaction(
+	sys::state& state,
+	dcon::nation_id s,
+	dcon::commodity_id c
+) {
+	auto d = demand(state, s, c);
+	auto con = consumption(state, s, c);
+	if(d == 0.f) {
+		return 0.f;
+	}
+	return con / d;
+}
+float demand_satisfaction(
+	sys::state& state,
+	dcon::commodity_id c
+) {
+	auto d = demand(state, c);
+	auto con = consumption(state, c);
+	if(d == 0.f) {
+		return 0.f;
+	}
+	return con / d;
 }
 
-void register_intermediate_demand(sys::state& state, dcon::nation_id n, dcon::commodity_id commodity_type, float amount, economy_reason reason) {
-	register_demand(state, n, commodity_type, amount, reason);
-	state.world.nation_get_intermediate_demand(n, commodity_type) += amount;
 
-	float price = state.world.commodity_get_current_price(commodity_type);
-	float sat = state.world.nation_get_demand_satisfaction(n, commodity_type);
+float average_capitalists_luxury_cost(
+	sys::state& state,
+	dcon::nation_id s
+) {
+	auto total = 0.f;
+	auto count = 0.f;
+	state.world.nation_for_each_state_ownership(s, [&](auto soid) {
+		auto sid = state.world.state_ownership_get_state(soid);
+		auto market = state.world.state_instance_get_market_from_local_market(sid);
+
+		auto local_count = state.world.state_instance_get_demographics(
+			sid,
+			demographics::to_key(state, state.culture_definitions.capitalists)
+		);
+
+		total += state.world.market_get_luxury_needs_costs(
+			market, state.culture_definitions.capitalists
+		) * local_count / state.defines.alice_needs_scaling_factor;
+		count += local_count;
+	});
 
-	state.world.nation_get_gdp(n) -= amount * price * sat;
+	return total / count;
 }
 
-// it's registered as a demand separately
-void register_construction_demand(sys::state& state, dcon::nation_id n, dcon::commodity_id commodity_type, float amount) {
-	state.world.nation_get_construction_demand(n, commodity_type) += amount;
+float export_volume(
+	sys::state& state,
+	dcon::market_id s,
+	dcon::commodity_id c
+){
+	return state.world.market_get_export(s, c);
+}
+float export_volume(
+	sys::state& state,
+	dcon::nation_id s,
+	dcon::commodity_id c
+){
+	float total = 0.f;
+	state.world.nation_for_each_state_ownership(s, [&](auto soid) {
+		auto sid = state.world.state_ownership_get_state(soid);
+		auto market = state.world.state_instance_get_market_from_local_market(sid);
+		total += state.world.market_get_export(market, c);
+	});
+	return total;
 }
 
-void register_domestic_supply(sys::state& state, dcon::nation_id n, dcon::commodity_id commodity_type, float amount, economy_reason reason) {
-	state.world.nation_get_domestic_market_pool(n, commodity_type) += amount;
-	state.world.nation_get_gdp(n) += amount * state.world.commodity_get_current_price(commodity_type);
+
+trade_volume_data_detailed export_volume_detailed(
+	sys::state& state,
+	dcon::nation_id s,
+	dcon::commodity_id c
+) {
+	static auto per_nation_data = state.world.nation_make_vectorizable_float_buffer();
```

Comment:
So, this may not work the way you want. By making it static, the buffer is only created once, which means that it will never grow if more nations are added. Now, currently, that probably isn't an issue since we create one nation per tag and just leave it like that, so in practice the the number nations are fixed. However, considering that this is a UI function (I assume) I would just remove the `static` here, paying the cost of allocating it once per call and making it more flexible for the future. On the other hand, this function is wedded to there being a fixed number of nations anyways as my next comment will explain ...